lib/addressable/idna.rb

```diff
@@ -16,11 +16,15 @@
 #    limitations under the License.
 #++
 
-
 begin
-  require "addressable/idna/native"
+  require "addressable/idna/native2"
 rescue LoadError
-  # libidn or the idn gem was not available, fall back on a pure-Ruby
-  # implementation...
-  require "addressable/idna/pure"
-end
+  # libidn2 or the ffi gem was not available, fall back on libidn1
+  begin
+    require "addressable/idna/native"
+  rescue LoadError
+    # libidn or the idn gem was not available, fall back on a pure-Ruby
+    # implementation...
+    require "addressable/idna/pure"
+  end
+end
```

Comment:
I like your suggestions. This has been (and still is) a bit of a gotcha (the tests have `IDNA_MODE` to control it). Sounds like the improvements needs to be introduced in a major version though?

Reply:
This wasn't as easy as I hoped because:
- `Addressable::IDNA = Addressable::IDNA::Native2` is not possible, if we re-assign `Addressable::IDNA` then `Addressable::IDNA::Native2` is no longer accessible because it was INSIDE the previous module we just shadowed by re-assigning. We can workaround this by changing the name of one or the other, but anyway:
- Re-assigning `Addressable::IDNA` is also not something Ruby likes simply because it's considered a CONSTANT so gives warnings when we change it (not a big deal but we can't recommend this as a way to select the backend if it generates warnings)

In the end I implented it another way in 62a136576446912fddde671ba41088df10ee9104:
- I set the backend in a class variable instead of a constant e.g: `Addressable::IDNA.backend = Addressable::IDNA::Libidn2`
- We can easily check the backend with `Addressable::IDNA.backend.name` (used it in profile:memory as an example)
- In order not to break compatibility with people who may be using `Addressable::IDNA.to_ascii` and `Addressable::IDNA.to_unicode`, I defined them as simple method in the IDNA module (delegating to the backend).
- I also changed the module/file names as discussed below at the same time (`libidn2`, `libidn1` and `pure`)

This has a couple side benefits:
- The common code like (deprecated) `unicode_normalize_kc` method can now be inside this `Addressable::IDNA` module instead of being duplicated and cluturing the backend modules.
- The public API is much clearer now (`to_ascii` and `to_unicode`) and other private or public method and constants in the backend are NOT accessible from `Addressable::IDNA`, reducing this kind of problem in the future. (people who want to hack can still call them from the backend directly, but in that case it's their problem).

Some downsides I see:
- The git history is a bit lost for these files, it'll show this comit because I namespaces the module. As the `pure` one which is the biggest is also gonna change a lot in #491, maybe not a big deal
- We can't really remove `IDNA_MODE` because even though it's simpler to switch in Ruby, we still need a way for the text matrix to communicate that choice to the code.

I updated the readme with some instructions and doc about this if you want a clearer picture.
Let me know what you think, and if you are not comfortable with this change we can also forget about it, no big deal.

Reply:
I like the way you took. Greate updates to the README.

I'm not too worried about the git history when we are making motivated changes, but there's always the [`.git-blame-ignore-revs`](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view) option to mitigate impact, but I think we are fine in this case.